text/0000-prelude-2021.md

```diff
@@ -0,0 +1,164 @@
+
+Without doing this, the `TryFrom`/`TryInto` traits remain less accessible than the infallible `From`/`Into` conversion traits, providing a disincentive to implement/use fallible conversions even where conversion operations are intrinsically fallible.
+
+## `FromIterator`
+
+The documentation for `FromIterator`'s `from_iter()` method currently reads:
+
+> `FromIterator::from_iter()` is rarely called explicitly, and is instead used through `Iterator::collect()` method. See `Iterator::collect()`'s documentation for more examples.
+
+However, it is reasonably common that type inferencing fails to infer the full type of the target type, in which case an explicit type annotation or turbofishing is needed (such as `iter.collect::<HashMap<_, _>>()` -- the type of the iterator item is available, so wildcards can be used for this). In these cases, `HashMap::from_iter(iter)` can be a more concise and readable way to spell this, which would be easier if the trait was in scope.
+
+## `std::{mem, fmt}`
+
+It is reasonably common to use functions from the `mem` module such as `replace()` or `swap()`. These method names are generic enough that it is useful to keep the `mem::` prefix explicitly. Providing a `mem` name as part of the prelude makes this pattern more obvious and, given the short module name, will not require any imports to make use of the items in this module.
+
+Explicit (non-derive) implementations of `Debug` and `Display` are relatively common and require a set of items from the `fmt` module (the trait itself, the `Formatter` type and the `fmt::Result` type).
+
+## `std::collections::{HashMap, HashSet}`
+
+Especially when coming from another language where hashmaps and sets are language built-ins with literals provided as part of the language, it can be a little surprising at first that these have to be imported before use. Their usage can be especially common in script-like code, for which effortless usage will be especially helpful.
+
+While a number of third-party hashmaps are in common use (most of which reexport or wrap the std `HashMap` with a different hasher), bringing the std `HashMap` into the prelude should not cause issues, since explicit imports will always override prelude imports.
+
+## `std::{{fs::File}, path::{Path, PathBuf}}`
+
+Are very commonly used, not least of which in script-like CLI tools.
+
+## `std::sync::{Arc, Mutex, RwLock}`
+
+Help write threaded code. A potential concern is that async code may want to use async variants of these, although even then the [sync Mutex](https://docs.rs/tokio/1.1.1/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use) may provide better performance.
+
+Use of alternative implementations (in particular from parking_lot) is fairly common.
```

Comment:
I think these are a bad choice because of the concern you raised. It seems like it will add a lot of confusion as to whether you're actually using an async mutex or not.

Reply:
Yes, also there have been discussion of adding another `Mutex`/`RwLock` type to libstd, for example a non-poisoning version. Given these concerns it seems like a dubious choice.

That said, I'm somewhat in favor of `Arc`.

Reply:
It's not uncommon to use crates like `archery` to generalize over `Arc` and `Rc`, so `Arc` would also be a poor fit for prelude.

Reply:
> It's not uncommon to use crates like archery to generalize over Arc and Rc, so Arc would also be a poor fit for prelude.

By itself, I don't find that a very strong argument. (Personally I also haven't seen any `Arc` alternatives in the wild -- archery doesn't seem that popular.)

Reply:
`servo_arc` and `triomphe` both exist. I also have my own `arcstr` crate (admittedly, not generic and not colliding).

That said yes, I don't find the existence of `archery` that compelling, but I also write a lot of multithreaded code and dont find reaching into std::sync that much of a big deal for any of these...

Reply:
As something of a devils advocate, `Arc<T>` tends to show up in function and type signatures more often than `Mutex<T>` does, which often IME is either in a static, or abstracted behind some other API. And because of this, it might have a stronger argument for inclusion.

That said, this all could be particular to my style of writing code.

Reply:
Although different, `(tokio|async_std)::sync::Mutex` <-> `std::sync::Mutex` prefix usage reminds me the same current prefix usage of `some_lib_crate::Result` and `std::result::Result`.

Reply:
It feels different. some_lib_crate::Result is still a std::result::Result. Also, the common pattern now is for lib crates to do

```rust
pub type Result<T, E = MyError> = std::result::Result<T, E>;
```

Allowing them to be used as normal `Result`s. I think the potential for libstd defining a different mutex (and in the discussion deprecating the std::sync one was even brought up) is more of a concern though.

Reply:
I agree with the rationale that Rc and Arc (and probably things like RefCell too) are likely a better fit than Mutex because while things like the archery crate exist, counters are comparatively fundamental memory-management constructs that interact much more directly with Rust's logic as a language, whereas Mutex is a way to manage locking and there's far more potential diversity in such, even std has Mutex, RwLock, Barrier, Atomic integers... etc... and all those can get wrapped in Arcs.

Reply:
I've moved `Mutex` and `RwLock` into the alternatives section for now; I especially find the discussion about non-poisoning alternatives to these a compelling argument to at least hold back from including these now.